File: cn/istarx/java/NO976_LargestPerimeterTriangle.java
package cn.istarx.java;

public class NO976_LargestPerimeterTriangle {
    public static void main(String[] args) {
        int[] A = new int[]{1, 1, 2};
        System.out.println(largestPerimeter(A));
    }

    private static int largestPerimeter(int[] A) {
        quickSort(A, 0, A.length - 1); // increase
        int numCount = A.length;
        for (int index = numCount - 3; index >= 0; index--) {
            if (A[index] + A[index + 1] > A[index + 2]) {
                return A[index] + A[index + 1] + A[index + 2];
            }
        }
        return 0;
    }

    private static void quickSort(int[] nums, int start, int end) {
        if (start >= end) {
            return;
        }
        int pivot = partition(nums, start, end);
        quickSort(nums, start, pivot - 1);
        quickSort(nums, pivot + 1, end);
    }

    private static int partition(int[] nums, int start, int end) {
        int pivot = nums[end];
        int i = start;
        for (int j = start; j < end; j++) {
            if (nums[j] < pivot) {
                swap(nums, i, j);
                i++;
            }
        }
        swap(nums, i, end);
        return i;
    }

    private static void swap(int[] nums, int i, int j) {
        int tmp = nums[j];
        nums[j] = nums[i];
        nums[i] = tmp;
    }
}

File: src/cn/istarx/python/NO976_LargestPerimeterTriangle.py
class Solution:
    def largestPerimeter(self, A: list) -> int:
        self.quick_sort(A, 0, len(A) - 1)
        for i in range(len(A) - 3, -1, -1): # -1 
            if A[i] + A[i + 1] > A[i + 2]:
                return A[i] + A[i + 1] + A[i + 2]
        return 0

    def quick_sort(self, A: list, start: int, end: int) -> None:
        if start >= end:
            return
        pivot = self.partition(A, start, end)
        self.quick_sort(A, start, pivot - 1)
        self.quick_sort(A, pivot + 1, end)

    @staticmethod
    def partition(A: list, start: int, end: int) -> int:
        pivot = A[end]
        i = start
        for j in range(start, end, 1):
            if A[j] < pivot:
                A[i], A[j] = A[j], A[i]
                i = i + 1
        A[i], A[end] = A[end], A[i]
        return i


if __name__ == '__main__':
    print(str(Solution().largestPerimeter([2, 1, 2])))
